### PR TITLE
Eliminate more primitive types from `balanceTx` type signature.

### PR DESCRIPTION
**Pull request by jonathanknowles**
## Issue

ADP-3157 (follow-on from #4130)

## Summary

This PR replaces the use of `Cardano.Wallet.Primitive` types with equivalent ledger types in the type signature of `balanceTransaction`.

### CODE REVIEW

File: lib/balance-tx/lib/Cardano/Wallet/Shelley/Compatibility/Ledger.hs [hunk 1]
```diff
@@ -295,7 +296,10 @@ instance Convert Address (Ledger.Addr StandardCrypto) where
             [ "toLedger @Address: Invalid address:"
             , pretty (Address bytes)
             ]
-    toWallet = Address . Ledger.serialiseAddr
+    toWallet = toWalletAddress
+
+toWalletAddress :: Ledger.Addr StandardCrypto -> Address
```
Comment (Anviking) [inline lib/balance-tx/lib/Cardano/Wallet/Shelley/Compatibility/Ledger.hs:305]: What's wrong with using the existing `toWallet`?

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 2]
```diff
@@ -3224,8 +3224,8 @@ prop_balanceTransactionValid
                         ]
             Left (ErrBalanceTxAssetsInsufficient err) -> do
                 let missing = view #shortfall err
-                let missingCoin = view #coin missing == Coin 0
-                let missingTokens = view #tokens missing == mempty
+                let missingCoin = Value.coin missing == mempty
+                let missingTokens = Value.isAdaOnly missing
```
Comment (Anviking) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:3237]: Hm... doesn't it look like the original check (and new) is the opposite of what it should be?

**Review by Anviking — APPROVED** [review]

File: lib/balance-tx/lib/Cardano/Wallet/Shelley/Compatibility/Ledger.hs (hunk 1 above)
Comment (jonathanknowles) [reply lib/balance-tx/lib/Cardano/Wallet/Shelley/Compatibility/Ledger.hs:305]: > What's wrong with using the existing `toWallet`?

There's nothing wrong with `toWallet`, but in some situations using it requires an additional type application to disambiguate types. We could avoid this type application by changing the functional dependency to be bidirectional.

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs (hunk 2 above)
Comment (jonathanknowles) [reply lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:3237]: > Hm... doesn't it look like the original check (and new) is the opposite of what it should be?

Good point! I'll try to create another PR that improves the naming a bit.

**Pull request merged**